code/ui/InventorySelection.cs

```diff
@@ -105,42 +109,63 @@ public override void Tick()
         [Event.BuildInput]
         private void ProcessClientInventorySelectionInput(InputBuilder input)
         {
-            Inventory inventory = Local.Pawn.Inventory as Inventory;
-
-            if (inventory.Count() == 0)
+            if (_children == null || _children.Count == 0)
             {
                 return;
             }
 
-            int selectedInventoryIndex = SlotPressInput(input);
+            ICarriableItem activeCarriable = Local.Pawn.ActiveChild as ICarriableItem;
 
-            // no button pressed, but maybe scrolled
-            if (selectedInventoryIndex == 0)
+            int keyboardIndexPressed = GetKeyboardNumberPressed(input);
+            if (keyboardIndexPressed != 0)
             {
-                if (input.MouseWheel != 0)
-                {
-                    int nextSlot = inventory.List.IndexOf(Local.Pawn.ActiveChild) - input.MouseWheel;
-                    nextSlot = inventory.NormalizeSlotIndex(nextSlot, inventory.Count() - 1);
+                // Find all weapons of hold type that equals the number the user pressed on their keyboard.
+                List<InventorySlot> weaponsOfHoldType = _children.ConvertAll(p => p as InventorySlot).Where(
+                    (slot) => (int) slot.Carriable.HoldType == keyboardIndexPressed).ToList();
 
-                    input.ActiveChild = inventory.GetSlot(nextSlot);
+                // Check to see if a user is currently holding a weapon which is of the same hold type.
+                bool isActiveCarriableOfWeaponType = weaponsOfHoldType.Any((slot) =>
+                    slot.Carriable.Name == activeCarriable?.Name);
 
-                    input.MouseWheel = 0;
+                if (activeCarriable == null || !isActiveCarriableOfWeaponType)
+                {
+                    // The user isn't holding a weapon, or is holding a weapon of a different hold type.
+                    // We can just select the first weapon.
+                    input.ActiveChild = weaponsOfHoldType.FirstOrDefault()?.Carriable as Entity;
+                }
+                else
+                {
+                    // The user is holding a weapon of the currently selected hold type.
+                    // Increment the index using GetNextWeaponIndex.
+                    int activeCarriableIndex = weaponsOfHoldType.FindIndex((slot) => slot.Carriable.Name == activeCarriable?.Name);
+                    input.ActiveChild = weaponsOfHoldType[GetNextWeaponIndex(activeCarriableIndex, weaponsOfHoldType.Count)].Carriable as Entity;
```

Comment:
You're looping two times through the entire list, couldn't that be optimized. That's why I don't like using Linq :D